File: 05-functions/my-code/202212/rpg.py
class RPGChar:

	def __init__(self, name: str, health: int, attack: int, defense: int,
	             x: str) -> None:
		self.name = name
		self.health = health
		self.attack = attack
		self.defense = defense
		self.x = x
		self.last_move = None
		self.special = False

	def soco(self, rpg: "RPG") -> None:
		rpg.seed_reroll()
		modifier = rpg.seed % 3
		damage = max(0, (self.attack - rpg.target.defense) * modifier)
		rpg.target.health -= damage
		rpg.event_print(rpg.target.name, "sofreu", damage, "dano")

	def arremesso_de_facas(self, rpg: "RPG") -> None:
		rpg.seed_reroll()
		hit_count = rpg.seed % 6
		damage = 0
		for hit_index in range(1, hit_count + 1):
			damage += self.attack // (3 ** hit_index)
		rpg.target.health -= damage
		rpg.event_print(rpg.target.name, "sofreu", damage, "dano")

	def _atk_or_def_up(self, rpg: "RPG", bonus: int) -> None:
		if bonus % 2 == 1:
			self.attack += bonus
			rpg.event_print(self.name, "ganhou", bonus, "ataque")
		else:
			self.defense += bonus
			rpg.event_print(self.name, "ganhou", bonus, "defesa")

	def invocacao_de_fada(self, rpg: "RPG") -> None:
		rpg.seed_reroll()
		hp_up = rpg.seed % 100
		rpg.seed_reroll()
		rpg.event_print(self.name, "ganhou", hp_up, "vida")
		bonus = rpg.seed % 1024
		if bonus < 100:
			self._atk_or_def_up(rpg, bonus)
		if bonus > 1018:
			rpg.user.special = True

	def execute_move(self, rpg: "RPG") -> None:
		match (self.last_move):
			case "soco":
				self.soco(rpg)
			case "facas":
				self.arremesso_de_facas(rpg)
			case "fada":
				self.invocacao_de_fada(rpg)


class RPG:

	def __init__(self, player: "RPGChar", enemy: "RPGChar",
	             seed: int) -> None:
		self.player = player
		self.enemy = enemy
		self.seed = seed
		self.player_turn = True
		self.battle_ended = False

	def set_user_target(self) -> None:
		if self.player_turn:
			self.user = self.player
			self.target = self.enemy
		else:
			self.user = self.enemy
			self.target = self.player

	def seed_reroll(self) -> None:
		self.seed = (7 * self.seed + 111) % 1024
		print(f"MENSAGEM DEBUG - número gerado: {self.seed}")

	def event_print(self, target: str, action: str, value: int,
	                stat: str) -> None:
		print(f"{target} {action} {value} pontos de {stat}!")

	def print_secret_ending(self) -> None:
		print("O quê? A fada trouxe um monstro gigante com ela!")
		if (self.player_turn):
			print(f"{self.enemy.name} e o castelo foram",
			      "destruídos,")
			print(f"e {self.player.name} agora tem um novo pet.")
			print("FINAL SECRETO - PARABENS???")
		else:
			print(f"{self.player.name} foi",
			      f"derrotad{self.player.x}...")
		self.battle_ended = True

	def print_regular_ending(self) -> None:
		if (self.player_turn):
			print(f"{self.enemy.name} foi derrotad{self.enemy.x}!",
			      f"{self.player.name} venceu!")
			print("FIM - PARABENS")
		else:
			print(f"{self.player.name} foi",
			      f"derrotad{self.player.x}...")
		self.battle_ended = True

	def check_battle_end(self) -> None:
		if self.user.special:
			self.print_secret_ending()
		elif self.player.health <= 0 or self.enemy.health <= 0:
			self.print_regular_ending()

File: 05-functions/my-code/202212/rpg.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "rpg.h"

struct rpg_char * create_character(char *name, char x)
{
	int hp, atk, def;
	struct rpg_char *c;

	scanf("%d %d %d", &hp, &atk, &def);

	c = malloc(sizeof(struct rpg_char));
	strcpy(c->name, name);
	c->hp = hp;
	c->atk = atk;
	c->def = def;
	c->special = 0;
	c->x = x;

	return c;
}

struct rpg * create_rpg(struct rpg_char *player, struct rpg_char *enemy,
                        int seed) {
	struct rpg *g;

	g = malloc(sizeof(struct rpg));
	g->player = player;
	g->enemy = enemy;
	g->seed = seed;
	g->player_turn = 1;
	g->battle_ended = 0;

	return g;
}

void set_user_target(struct rpg *g)
{
	if (g->player_turn) {
		g->user = g->player;
		g->target = g->enemy;
	} else {
		g->user = g->enemy;
		g->target = g->player;
	}
}

void reroll_seed(struct rpg *g)
{
	g->seed = (7 * g->seed + 111) % 1024;
	printf("MENSAGEM DEBUG - número gerado: %d\n", g->seed);
}

void print_event(char *target, char *action, int value, char *stat)
{
	printf("%s %s %d pontos de %s!\n", target, action, value, stat);
}

void punch(struct rpg *g)
{
	int modifier, damage;

	reroll_seed(g);
	modifier = g->seed % 3;
	damage = (g->user->atk - g->target->def) * modifier;
	damage = damage > 0 ? damage : 0;
	g->target->hp -= damage;
	print_event(g->target->name, "sofreu", damage, "dano");
}

void throw_knives(struct rpg *g)
{
	int hit_count, damage, i;

	reroll_seed(g);
	hit_count = g->seed % 6;
	damage = 0;
	for (i = 1; i < hit_count + 1; i++)
		damage += g->user->atk / pow(3, i);
	g->target->hp -= damage;
	print_event(g->target->name, "sofreu", damage, "dano");
}

void raise_hp(struct rpg_char *user, int raise_value)
{
	user->hp += raise_value;
	print_event(user->name, "ganhou", raise_value, "vida");
}

void raise_atk(struct rpg_char *user, int raise_value)
{
	user->atk += raise_value;
	print_event(user->name, "ganhou", raise_value, "ataque");
}

void raise_def(struct rpg_char *user, int raise_value)
{
	user->def += raise_value;
	print_event(user->name, "ganhou", raise_value, "defesa");
}

void summon_fairy(struct rpg *g)
{
	int hp_up, bonus;

	reroll_seed(g);
	hp_up = g->seed % 100;
	reroll_seed(g);
	raise_hp(g->user, hp_up);
	bonus = g->seed % 1024;
	if (bonus < 100) {
		if (bonus % 2 == 1)
			raise_atk(g->user, bonus);
		else
			raise_def(g->user, bonus);
	} else if (bonus > 1018)
		g->user->special = 1;
}

void execute_move(struct rpg *g)
{
	if (strcmp(g->user->last_move, "soco") == 0)
		punch(g);
	else if (strcmp(g->user->last_move, "facas") == 0)
		throw_knives(g);
	else if (strcmp(g->user->last_move, "fada") == 0)
		summon_fairy(g);
}

void print_secret_ending(struct rpg *g)
{
	printf("O quê? A fada trouxe um monstro gigante com ela!\n");
	if (g->player_turn) {
		printf("%s e o castelo foram destruídos,\n", g->enemy->name);
		printf("e %s agora tem um novo pet.\n", g->player->name);
		printf("FINAL SECRETO - PARABENS???\n");
	} else
		printf("%s foi derrotad%c...\n", g->player->name, g->player->x);
	g->battle_ended = 1;
}

void print_regular_ending(struct rpg *g)
{
	if (g->player_turn) {
		printf("%s foi derrotad%c! ", g->enemy->name, g->enemy->x);
		printf("%s venceu!\n", g->player->name);
		printf("FIM - PARABENS\n");
	} else
		printf("%s foi derrotad%c...\n", g->player->name, g->player->x);
	g->battle_ended = 1;
}

void check_battle_end(struct rpg *g)
{
	if (g->user->special)
		return print_secret_ending(g);
	if (g->player->hp <= 0 || g->enemy->hp <= 0)
		return print_regular_ending(g);
}
